library/Vanilla/TokenSigningTrait.php

```diff
@@ -0,0 +1,199 @@
+        }
+
+        $checkToken = $this->signToken($token, $expires);
+        if (!hash_equals($checkToken, $accessToken)) {
+            return $this->tokenError('Your '.$tokenIdentifier.' has an invalid signature.', 401, $throw);
+        }
+
+        return true;
+    }
+
+    /**
+     * Sign a token generated with {@link randomToken()}.
+     *
+     * @param string $token The token to sign.
+     * @param mixed $expires The expiry date of the token.
+     * @return string
+     */
+    public function signToken($token, $expires = '2 months') {
+        if (empty($this->getSecret())) {
+            // This means something has been misconfigured. Throw a noisy error.
+            throw new \Exception("No secret to sign tokens with.", 500);
+        }
+
+        $str = 'va.'.$token.'.'.$this->encodeDate($expires);
+        $sig = self::base64urlEncode(hash_hmac('sha256', $str, $this->secret, true));
+
+        // Use a substring of the signature because we don't want the tokens to be too massive.
+        // The signature is only the first line of defence. The database is the final verification.
+        $result = $str.'.'.substr($sig, 0, 7);
+        return $result;
+    }
+
+    /**
+     * Optionally throw an exception.
+     *
+     * @param string $message The error message.
+     * @param int $code The error code.
+     * @param bool $throw Whether or not to throw an exception.
+     * @return bool Returns **false**.
+     * @throws Exception Throws an exception if {@link $throw} is true.
+     */
+    private function tokenError($message, $code = 401, $throw = false) {
+        if ($throw) {
+            throw new \Exception($message, $code);
+        }
+        return false;
+    }
+
+    /**
+     * Force a value into a timestamp.
+     *
+     * @param mixed $dt A timestamp or date string.
+     * @return false|int
+     */
+    private function toTimestamp($dt) {
+        if (is_numeric($dt)) {
+            return (int)$dt;
+        } elseif ($ts = strtotime($dt)) {
+            return $ts;
+        }
+        return null;
+    }
+
+    /**
+     * Base 64 encode a date.
+     *
+     * @param mixed $dt A timestamp or date string.
+     * @return string Returns the encoded date.
+     */
+    private function encodeDate($dt) {
+        $timestamp = $this->toTimestamp($dt);
+        $result = self::base64urlEncode(pack('I', $timestamp));
+        return $result;
+    }
+
+    /**
+     * Base 64 decode a date.
+     *
+     * @param string $str An encoded date.
+     * @return int Returns a timestamp.
+     */
+    private function decodeDate($str) {
+        $arr = unpack('I*', self::base64urlDecode($str));
+        if (empty($arr[1]) || !is_int($arr[1])) {
+            return null;
+        }
+        return $arr[1];
+    }
+
+    /**
+     * Generate and sign a token.
+     *
+     * @param string $expires When the token expires.
+     * @return string
+     */
+    public function randomSignedToken($expires = '2 months') {
```

Comment:
Why would this be better in here, instead of staying in `AccessTokenModel`? It looks like it might only be used there.

Reply:
I use this in the TokenSigningTraitTest